File: kielo_shared/gcs_utils.py
"""
Kielo GCS utilities — language-agnostic helpers derived from kielo-shared/gcs.json.

Usage:
    from kielo_shared.gcs_utils import ensure_alt_media, build_object_url, is_storage_api_path, BUCKETS
"""

import json
import os
from pathlib import Path
from urllib.parse import urlparse, urlunparse, quote

_CONFIG_PATH = Path(__file__).parent / "gcs.json"
_CONFIG = json.loads(_CONFIG_PATH.read_text())

STORAGE_API_PATH: str = _CONFIG["storageAPIPath"]
UPLOAD_API_PATH: str = _CONFIG["uploadAPIPath"]
ALT_MEDIA_PARAM: str = _CONFIG["altMediaParam"]
DEFAULT_EMULATOR_PORT: int = _CONFIG["defaultEmulatorPort"]
BUCKETS: dict[str, str] = _CONFIG["buckets"]
ENV_VARS: dict[str, str] = _CONFIG["envVars"]


def is_storage_api_path(url_path: str) -> bool:
    """Check if a URL path is a GCS storage API path."""
    return url_path.startswith(STORAGE_API_PATH) or url_path.startswith(UPLOAD_API_PATH)


def ensure_alt_media(url: str) -> str:
    """Append ?alt=media to a URL if not already present."""
    if ALT_MEDIA_PARAM in url:
        return url
    separator = "&" if "?" in url else "?"
    return f"{url}{separator}{ALT_MEDIA_PARAM}"


def build_object_url(base: str, bucket: str, object_path: str) -> str:
    """Build a GCS object URL: {base}/storage/v1/b/{bucket}/o/{encoded_object}"""
    base = base.rstrip("/")
    if base.endswith("/storage/v1"):
        base = base[: -len("/storage/v1")]
    if base.endswith("/storage"):
        base = base[: -len("/storage")]
    encoded = quote(object_path, safe="")
    return f"{base}{STORAGE_API_PATH}{bucket}/o/{encoded}"


def build_object_fetch_url(base: str, bucket: str, object_path: str) -> str:
    """Build a GCS object URL with ?alt=media for fetching content."""
    return build_object_url(base, bucket, object_path) + f"?{ALT_MEDIA_PARAM}"


def get_emulator_host() -> str:
    """Get the normalized emulator host from STORAGE_EMULATOR_HOST env var."""
    raw = os.getenv(ENV_VARS["emulatorHost"], "").strip()
    if not raw:
        return ""
    if "://" not in raw:
        raw = f"http://{raw}"
    parsed = urlparse(raw)
    return f"{parsed.scheme}://{parsed.hostname}:{parsed.port or DEFAULT_EMULATOR_PORT}"


def get_internal_emulator_host() -> str:
    """Get the internal Docker emulator host, rewriting localhost to gcs-emulator."""
    host = get_emulator_host()
    if not host:
        return ""
    parsed = urlparse(host)
    hostname = parsed.hostname or ""
    if hostname in ("localhost", "127.0.0.1", "::1"):
        return f"{parsed.scheme}://gcs-emulator:{parsed.port or DEFAULT_EMULATOR_PORT}"
    return host


def normalize_internal_storage_url(url: str) -> str:
    """Rewrite external emulator URLs to internal Docker hostname."""
    parsed = urlparse(url)
    if not is_storage_api_path(parsed.path):
        return url
    internal = get_internal_emulator_host()
    if not internal:
        return url
    internal_parsed = urlparse(internal)
    if parsed.hostname == internal_parsed.hostname:
        return url
    return urlunparse(parsed._replace(
        scheme=internal_parsed.scheme,
        netloc=f"{internal_parsed.hostname}:{internal_parsed.port or DEFAULT_EMULATOR_PORT}",
    ))

File: ts/gcs_utils.ts
/**
 * Kielo GCS utilities — language-agnostic helpers derived from kielo-shared/gcs.json.
 *
 * Usage:
 *   import { ensureAltMedia, buildObjectUrl, isStorageApiPath, BUCKETS } from "@kielo/shared/gcs_utils";
 */

import config from "../gcs.json";

export const STORAGE_API_PATH: string = config.storageAPIPath;
export const UPLOAD_API_PATH: string = config.uploadAPIPath;
export const ALT_MEDIA_PARAM: string = config.altMediaParam;
export const DEFAULT_EMULATOR_PORT: number = config.defaultEmulatorPort;
export const BUCKETS = config.buckets;
export const ENV_VARS = config.envVars;

/** Check if a URL path is a GCS storage API path. */
export function isStorageApiPath(urlPath: string): boolean {
  return urlPath.startsWith(STORAGE_API_PATH) || urlPath.startsWith(UPLOAD_API_PATH);
}

/**
 * Append ?alt=media to GCS JSON-API object URLs so they return content.
 *
 * Only JSON-API paths (/storage/v1/... or /upload/storage/v1/...) need the
 * parameter. Signed URLs must pass through untouched: a V4 signature covers
 * the whole query string, so appending anything invalidates it and GCS
 * returns 403 SignatureDoesNotMatch — this broke every signed hero image in
 * the notification inbox (prod + emulator, 2026-08-08; same defect class as
 * the 2026-08-06 transcription 403 fixed in the Python twin of this
 * helper). Public XML-API/CDN URLs don't need the parameter either.
 * Skips non-remote schemes (data:, file:, ph:, blob:) and no-ops if the
 * param is already there.
 */
export function ensureAltMedia(url: string): string {
  if (
    !url ||
    url.startsWith("data:") ||
    url.startsWith("file:") ||
    url.startsWith("ph:") ||
    url.startsWith("blob:") ||
    url.includes(ALT_MEDIA_PARAM)
  ) {
    return url;
  }
  if (url.includes("X-Goog-Signature") || url.includes("X-Goog-Algorithm")) {
    return url;
  }

  try {
    const parsed = new URL(url);
    if (!isStorageApiPath(parsed.pathname)) {
      return url;
    }
    if (parsed.searchParams.get("alt") !== "media") {
      parsed.searchParams.set("alt", "media");
    }
    return parsed.toString();
  } catch {
    // Relative/unparseable URLs: only the JSON-API path shape needs the
    // parameter; anything else passes through unchanged.
    if (!isStorageApiPath(url)) {
      return url;
    }
    const separator = url.includes("?") ? "&" : "?";
    return `${url}${separator}${ALT_MEDIA_PARAM}`;
  }
}

/** Build a GCS object URL: {base}/storage/v1/b/{bucket}/o/{encodedObject} */
export function buildObjectUrl(base: string, bucket: string, objectPath: string): string {
  let b = base.replace(/\/+$/, "");
  if (b.endsWith("/storage/v1")) b = b.slice(0, -"/storage/v1".length);
  if (b.endsWith("/storage")) b = b.slice(0, -"/storage".length);
  return `${b}${STORAGE_API_PATH}${bucket}/o/${encodeURIComponent(objectPath)}`;
}

/** Build a GCS object URL with ?alt=media for fetching content. */
export function buildObjectFetchUrl(base: string, bucket: string, objectPath: string): string {
  return `${buildObjectUrl(base, bucket, objectPath)}?${ALT_MEDIA_PARAM}`;
}
